linux/src/completion_port_linux.c

```diff
@@ -0,0 +1,434 @@
+                    ANNOTATE_HAPPENS_BEFORE(epoll_thread_data);
+                #endif
+                // Codes_COMPLETION_PORT_LINUX_11_022: [ completion_port_add shall add the EPOLL_THREAD_DATA object to a list for later removal. ]
+                if (add_thread_data_to_list(completion_port, epoll_thread_data) != 0)
+                {
+                    LogError("failure adding data to list");
+                }
+                else
+                {
+                    // Codes_COMPLETION_PORT_LINUX_11_023: [ completion_port_add shall add the socket in the epoll system by calling epoll_ctl with EPOLL_CTL_MOD along with the epoll_op variable. ]
+                    if (epoll_ctl(completion_port->epoll, EPOLL_CTL_MOD, epoll_thread_data->socket, &ev) < 0)
+                    {
+                        if (errno == ENOENT)
+                        {
+                            // Codes_COMPLETION_PORT_LINUX_11_024: [ If the epoll_ctl call fails with ENOENT, completion_port_add shall call epoll_ctl again with EPOLL_CTL_ADD. ]
+                            if (epoll_ctl(completion_port->epoll, EPOLL_CTL_ADD, epoll_thread_data->socket, &ev) < 0)
+                            {
+                                LogErrorNo("failure with epoll_ctl EPOLL_CTL_ADD");
+                            }
+                            else
+                            {
+                                result = 0;
+                                (void)interlocked_decrement(&completion_port->pending_calls);
+                                wake_by_address_single(&completion_port->pending_calls);
+                                goto all_ok;
+                            }
+                        }
+                        else
+                        {
+                            LogErrorNo("failure with epoll_ctl EPOLL_CTL_MOD");
+                        }
+                    }
+                    else
+                    {
+                        // Codes_COMPLETION_PORT_LINUX_11_026: [ On success, completion_port_add shall return 0. ]
+                        result = 0;
+                        // Codes_COMPLETION_PORT_LINUX_11_025: [ completion_port_add shall decrement the ongoing call count value to unblock close. ]
+                        (void)interlocked_decrement(&completion_port->pending_calls);
+                        wake_by_address_single(&completion_port->pending_calls);
+                        goto all_ok;
+                    }
+                    // Remove
+                    remove_thread_data_from_list(completion_port, epoll_thread_data);
+                }
+                free(epoll_thread_data);
+            }
+            // Codes_COMPLETION_PORT_LINUX_11_027: [ If any error occurs, completion_port_add shall fail and return a non-zero value. ]
+            result = MU_FAILURE;
+            (void)interlocked_decrement(&completion_port->pending_calls);
+            wake_by_address_single(&completion_port->pending_calls);
```

Comment:
if you move "all_ok:" up here then you don't need to repeat these 2 lines 3 times.